### DISCUSSION TITLE
unable to use podman binding in go program

### DISCUSSION THREAD

**Discussion by amokkara** [containers/podman#15685] [General]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

unable to use podman binding in go program

**Steps to reproduce the issue:**

i am facing below error when i try to build my go program which uses go binding

**# github.com/containers/image/v5/manifest
/root/go/pkg/mod/github.com/containers/image/v5@v5.16.0/manifest/oci.go:44:244: undefined: v1.MediaTypeImageLayerNonDistributableZstd
/root/go/pkg/mod/github.com/containers/image/v5@v5.16.0/manifest/oci.go:44: undefined: v1.MediaTypeImageLayerZstd
/root/go/pkg/mod/github.com/containers/image/v5@v5.16.0/manifest/oci.go:101:49: undefined: v1.MediaTypeImageLayerNonDistributableZstd
/root/go/pkg/mod/github.com/containers/image/v5@v5.16.0/manifest/oci.go:106:49: undefined: v1.MediaTypeImageLayerZstd**

i have these 2 imports 
"github.com/containers/podman/v2/pkg/bindings"
"github.com/containers/podman/v2/pkg/bindings/images"
	
i am using go v1.17.3	

**Describe the results you received:**

**Describe the results you expected:**

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `podman version`:**

```
Version:      3.4.2
API Version:  3.4.2
Go Version:   go1.15.2
Built:        Thu Jan  1 05:30:00 1970
OS/Arch:      linux/amd64
```

**Output of `podman info`:**

```
host:
  arch: amd64
  buildahVersion: 1.23.1
  cgroupControllers:
  - cpuset
  - cpu
  - cpuacct
  - blkio
  - memory
  - devices
  - freezer
  - net_cls
  - perf_event
  - net_prio
  - hugetlb
  - pids
  - rdma
  cgroupManager: systemd
  cgroupVersion: v1
  conmon:
    package: 'conmon: /usr/libexec/podman/conmon'
    path: /usr/libexec/podman/conmon
    version: 'conmon version 2.1.2, commit: '
  cpus: 4
  distribution:
    codename: focal
    distribution: ubuntu
    version: "20.04"
  eventLogger: journald
  hostname: scspo2665413001
  idMappings:
    gidmap: null
    uidmap: null
  kernel: 5.4.0-37-generic
  linkmode: dynamic
  logDriver: journald
  memFree: 271413248
  memTotal: 16791576576
  ociRuntime:
    name: crun
    package: 'crun: /usr/bin/crun'
    path: /usr/bin/crun
    version: |-
      crun version UNKNOWN
      commit: ea1fe3938eefa14eb707f1d22adff4db670645d6
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +YAJL
  os: linux
  remoteSocket:
    exists: true
    path: /run/podman/podman.sock
  security:
    apparmorEnabled: true
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: false
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: false
  serviceIsRemote: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: 'slirp4netns: /usr/bin/slirp4netns'
    version: |-
      slirp4netns version 1.1.8
      commit: unknown
      libslirp: 4.3.1-git
      SLIRP_CONFIG_VERSION_MAX: 3
      libseccomp: 2.4.3
  swapFree: 0
  swapTotal: 0
  uptime: 403h 4m 58.92s (Approximately 16.79 days)
plugins:
  log:
  - k8s-file
  - none
  - journald
  network:
  - bridge
  - macvlan
  volume:
  - local
registries:
  hub.docker.netapp.com:
    Blocked: false
    Insecure: true
    Location: hub.docker.netapp.com
    MirrorByDigestOnly: false
    Mirrors: []
    Prefix: hub.docker.netapp.com
  search:
  - hub.docker.netapp.com
store:
  configFile: /etc/containers/storage.conf
  containerStore:
    number: 0
    paused: 0
    running: 0
    stopped: 0
  graphDriverName: overlay
  graphOptions:
    overlay.mountopt: nodev,metacopy=on
  graphRoot: /var/lib/containers/storage
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "true"
  imageStore:
    number: 0
  runRoot: /run/containers/storage
  volumePath: /var/lib/containers/storage/volumes
version:
  APIVersion: 3.4.2
  Built: 0
  BuiltTime: Thu Jan  1 05:30:00 1970
  GitCommit: ""
  GoVersion: go1.15.2
  OsArch: linux/amd64
  Version: 3.4.2
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
(paste your output here)
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide? (https://github.com/containers/podman/blob/main/troubleshooting.md)**

Yes/No

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Comment by giuseppe** [comment]
how have you configured your go modules?

**Reply by amokkara** [reply]
@giuseppe  can we use podman binding 4.0.0 (v4) on podman server 3.4.7?

i keep getting error stating "server version much lower than client version"..something like that

**Reply by rhatdan** [reply]
No you can not use client and server from different Major versions.

**Reply by amokkara** [reply]
@giuseppe @rhatdan  Thanks. i think the issue was with my go module set up. its resolved now, although I have a doubt regarding dependencies.

https://podman.io/getting-started/installation#build-and-run-dependencies

there are no dependencies listed for alpine distribution. I'm trying to compile on alpine. Is there any way?

**Reply by rhatdan** [reply]
Well once you figure it out, you can update the docs.  :^)

**Reply by amokkara** [reply]
github.com/containers/podman/v3/pkg/rootless
/usr/lib/gcc/x86_64-alpine-linux-musl/10.3.1/../../../../x86_64-alpine-linux-musl/bin/ld: $WORK/b235/_x003.o: in function `cleanup_closep':
/root/go/pkg/mod/github.com/containers/podman/v3@v3.4.7/pkg/rootless/rootless_linux.c:38: undefined reference to `TEMP_FAILURE_RETRY'
/usr/lib/gcc/x86_64-alpine-linux-musl/10.3.1/../../../../x86_64-alpine-linux-musl/bin/ld: /root/go/pkg/mod/github.com/containers/podman/v3@v3.4.7/pkg/rootless/rootless_linux.c:38: undefined reference to `TEMP_FAILURE_RETRY'
/usr/lib/gcc/x86_64-alpine-linux-musl/10.3.1/../../../../x86_64-alpine-linux-musl/bin/ld: /root/go/pkg/mod/github.com/containers/podman/v3@v3.4.7/pkg/rootless/rootless_linux.c:38: undefined reference to `TEMP_FAILURE_RETRY'
/usr/lib/gcc/x86_64-alpine-linux-musl/10.3.1/../../../../x86_64-alpine-linux-musl/bin/ld: /root/go/pkg/mod/github.com/containers/podman/v3@v3.4.7/pkg/rootless/rootless_linux.c:38: undefined reference to `TEMP_FAILURE_RETRY'
/usr/lib/gcc/x86_64-alpine-linux-musl/10.3.1/../../../../x86_64-alpine-linux-musl/bin/ld: /root/go/pkg/mod/github.com/containers/podman/v3@v3.4.7/pkg/rootless/rootless_linux.c:38: undefined reference to `TEMP_FAILURE_RETRY'
collect2: error: ld returned 1 exit status
 github.com/containers/podman/v3/pkg/rootless
rootless_linux.c: In function 'cleanup_closep':
rootless_linux.c:38:5: warning: implicit declaration of function 'TEMP_FAILURE_RETRY' [-Wimplicit-function-declaration]
   38 |     TEMP_FAILURE_RETRY (close (*pp));
      |     ^~~~~~~~~~~~~~~~~~

**Reply by amokkara** [reply]
i am getting this error in alpine after installing the dependencies required. 
any thoughts?

**Reply by amokkara** [reply]
@rhatdan  atleast I dont see any dependencies missing error now. but this seems to be issue with code?!
could you take a look please

**Reply by amokkara** [reply]
looks similar to  - https://github.com/flatpak/xdg-dbus-proxy/issues/19

**Reply by amokkara** [reply]
> Well once you figure it out, you can update the docs. :^)

@rhatdan I have a list of dependencies that need to be installed for alpine. I was able to compile now by installing them. How can i update the docs?

**Reply by thediveo** [reply]
@amokkara can you please be so kind and post the list of packages that you needed to install for alpine here? I'm at this very moment trying to get a v3 podman REST API client build in a container image build stage using an alpine-based go image. I'm currently stuck with the undefined reference to `TEMP_FAILURE_RETRY` that blocks me. Not that it is required for the client, but this monolithic podman repo is a P.I.T.A...

**Reply by rhatdan** [reply]
In upstream that is defined. in podman/pkg/rooless_linux.c

#ifndef TEMP_FAILURE_RETRY
#define TEMP_FAILURE_RETRY(expression) \
  (__extension__                                                              \
    ({ long int __result;                                                     \
       do __result = (long int) (expression);                                 \
       while (__result == -1L && errno == EINTR);                             \
       __result; }))
#endif

**Reply by rhatdan** [reply]
On Fedora it is defined in the system file unistd.c

$ grep -r  TEMP_FAILURE_RETRY  /usr/include/
/usr/include/unistd.h:# define TEMP_FAILURE_RETRY(expression) \

**Reply by rhatdan** [reply]
@amokkara open a pr in github.com/containers/podman/podman.io or github.com/containers/podman/podman. depending on where instructions on building podman are defined.

**Reply by amokkara** [reply]
@thediveo here are the dependencies i installed

apk add btrfs-progs && apk add go-md2man && apk add iptables && apk add libassuan-dev && apk add btrfs-progs-dev && apk add libc-dev && apk add device-mapper-libs && apk add lvm2-dev && apk add device-mapper && glib-dev && apk add gpgme-dev && apk add libgpg-error-dev && apk add protobuf-dev && apk add protobuf-c-dev && apk add libseccomp-dev && apk add libselinux-dev && apk add pc:libsystemd && apk add elogind-dev && apk add runc && apk add shadow-uidmap && apk add pkgconfig

**Reply by amokkara** [reply]
@rhatdan im trying to cross compile my program to generate mac os binary but im getting error 

GOOS=darwin GOARCH=amd64 go build -o podman-poc   -> 
it gives below error:
    ```/root/go/pkg/mod/github.com/containers/image/v5@v5.17.0/signature/mechanism_gpgme.go:127:20: undefined: gpgme.NewDataBytes
/root/go/pkg/mod/github.com/containers/image/v5@v5.17.0/signature/mechanism_gpgme.go:132:18: undefined: gpgme.NewDataWriter
/root/go/pkg/mod/github.com/containers/image/v5@v5.17.0/signature/mechanism_gpgme.go:136:25: undefined: gpgme.Key
/root/go/pkg/mod/github.com/containers/image/v5@v5.17.0/signature/mechanism_gpgme.go:136:61: undefined: gpgme.SigModeNormal```


im able to generate linux binary without any issue though. im using ubuntu 20.04, golang - 1.17.1

I already opened a bug for this - https://github.com/containers/podman/issues/15733

but just wanted to check if this is any known issue. This is happening with podman/binding/v3 and v4 as well. any idea?

**Reply by thediveo** [reply]
> In upstream that is defined. in podman/pkg/rooless_linux.c
> 
> #ifndef TEMP_FAILURE_RETRY
> #define TEMP_FAILURE_RETRY(expression) \
>   (__extension__                                                              \
>     ({ long int __result;                                                     \
>        do __result = (long int) (expression);                                 \
>        while (__result == -1L && errno == EINTR);                             \
>        __result; }))
> #endif
> 

Unfortunately that's in the v4 branch as far as I can see, but not in the v3 branch.

**Reply by Luap99** [reply]
You do not need any c dependencies for the bindings if you use the correct remote build tags: https://github.com/containers/podman/blob/b9cbc0c09a68be7bb55048a494753c77c08bfbd3/Makefile#L45

**Reply by thediveo** [reply]
unfortunately, I'm not building using the `Makefile`, but importing podman only. While I've set some of the exclude build tags I don't see a way to exclude the offending rootless stuff?

**Reply by Luap99** [reply]
I thought the remote tag would disable it but maybe not. `CGO_ENABLED=0` should make sure no c code is compiled in.

**Reply by thediveo** [reply]
...`CGO_ENABLED=0` unfortunately isn't an option when my other parts need it...

**Reply by thediveo** [reply]
Okay, so in order to build the podman REST API client on Alpine musl, in the container build...
1. `apk add git patch linux-headers gpgme-dev libassuan-dev libgpg-error-dev`
1. create a Go modules workspace
2. check out podman @v3.4.7
3. patch `podman/pkg/rootless/rootless_linux.c`
4. copy my application sources into a second workspace module
5. use both podman and my app module in the workspace
6. go build my application using `-tags exclude_graphdriver_btrfs,exclude_graphdriver_devicemapper,libdm_no_deferred_remove`

The diff file:

```diff
--- rootless_linux.c    2022-09-13 17:00:00.185241165 +0200
+++ rootless_linux-main.c       2022-09-13 17:00:12.069395845 +0200
@@ -19,6 +19,15 @@
 #include <sys/select.h>
 #include <stdio.h>
 
+#ifndef TEMP_FAILURE_RETRY
+#define TEMP_FAILURE_RETRY(expression) \
+  (__extension__                                                              \
+    ({ long int __result;                                                     \
+       do __result = (long int) (expression);                                 \
+       while (__result == -1L && errno == EINTR);                             \
+       __result; }))
+#endif
+
 #define cleanup_free __attribute__ ((cleanup (cleanup_freep)))
 #define cleanup_close __attribute__ ((cleanup (cleanup_closep)))
 #define cleanup_dir __attribute__ ((cleanup (cleanup_dirp)))
@@ -72,15 +81,6 @@
   return rename (oldpath, newpath);
 }
 
-#ifndef TEMP_FAILURE_RETRY
-#define TEMP_FAILURE_RETRY(expression) \
-  (__extension__                                                              \
-    ({ long int __result;                                                     \
-       do __result = (long int) (expression);                                 \
-       while (__result == -1L && errno == EINTR);                             \
-       __result; }))
-#endif
-
 static const char *_max_user_namespaces = "/proc/sys/user/max_user_namespaces";
 static const char *_unprivileged_user_namespaces = "/proc/sys/kernel/unprivileged_userns_clone";
```

**Reply by giuseppe** [reply]
could you please open a PR?

**Reply by thediveo** [reply]
@giuseppe that would need to be against v3.4.7, so would you even accept it? v4 has moved the macro definition up accordingly; but it is v3 that still has the problems.

**Reply by mheon** [reply]
I have no issues merging it, but I make no promises we will ever do another release off the v3.4 branch.

**Reply by thediveo** [reply]
PR #15805

**Reply by thediveo** [reply]
for me, I can simply pin the latest commit on the v3.4 branch in my dependencies and then am fine with that, no more need to juggle with the Go modules workspace and patching sources on-the-fly.
